python/lsst/afw/coord/refraction.py

```diff
@@ -0,0 +1,284 @@
+from __future__ import absolute_import, division, print_function
+#
+# LSST Data Management System
+# Copyright 2008-2018 LSST/AURA.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+from astropy import units as u
+from astropy.units import cds
+import numpy as np
+
+from lsst.afw.geom import Angle
+from lsst.afw.coord import Observatory, Weather
+from lsst.afw.geom import degrees
+
+__all__ = ["refraction", "differentialRefraction"]
+
+lsstLat = -30.244639*degrees
+lsstLon = -70.749417*degrees
+lsstAlt = 2663.
+lsstTemperature = 20.*u.Celsius  # in degrees Celcius
+lsstHumidity = 10.  # in percent
+lsstPressure = 101325.*u.pascal  # 1 atmosphere.
```

Comment:
I totally agree with Paul here, but he doesn't make a suggestion.  I think that the only sensible solution is to add an observatory-specific object (singleton?) to the obs packages, which can  be imported or possibly be returned by the butler (cf. `camera`).  You'd then need to either pass this to the objects or import it via some camera-agnostic name.

Reply:
I agree in principle, and am comfortable removing these defaults. Both are typically stored in an exposures `VisitInfo`, but the problem I have been having is that the `weather` entry is often all nan's. The weather is only a second order effect for the DCR calculations I care about, so I do want the code to work even if that isn't present.

Reply:
@RobertLuptonTheGood @PaulPrice I've added a function that calculates reasonable default values if the weather parameters are `NAN`. Will this satisfy your concerns? One problem I see is that it won't catch parameters that are merely bad, and not `NAN`. In the Decam data I have, for example, the pressure is set to 1 atmosphere.

Reply:
I think the obs package should take responsibility for setting the correct values or reasonable defaults. That's the approach taken with `VisitInfo`.

Reply:
That sounds reasonable, but I would like the option of calculating expected atmospheric parameters. For example, the Decam data I mentioned above has non-`NAN` but bogus values in the `VisitInfo`, and anything read with `obs_lsstSim` has `NAN` values for humidity. What if I calculate default values if `weather` is not passed in at all (maybe log a warning), but raise an error if `weather` is passed in with invalid parameters?
My concern is that the weather conditions are not otherwise depended on, and it doesn't seem like it's usually a high priority to make sure they are present and correct.

Reply:
Couldn't you fix the `VisitInfo`s in those obs packages?

Reply:
What is there to fix in the case of `obs_decam`? It correctly reads the value from the input data correctly, but that was stored as 1 atmosphere instead of the real value. I don't want to have to manually fix the data, but if I use that value, then the refraction I calculate will be ~20% too high.

Reply:
It's the obs package's job to provide the correct values. If obs_decam isn't providing correct values, then I think that's the place to plug in your approximation.

Reply:
I think that would require a substantial change to the `obs_*` packages, and likely involves a lengthy discussion. For now, I would like to use the approximation in *all* cases, even when `weather` information is present. I would also create a new ticket to sanitize `weather` when it is read in, and hook it in to the refraction code at that point.

Reply:
My preference would be to do it properly now, but perhaps there are good reasons to cut corners. @jdswinbank can advise.

Reply:
Can we be more specific about what's actually being proposed here?

If it's simply adding some sort of `if` statement around [this](https://github.com/lsst/obs_decam/blob/921a630251e9741fb10e1cc7b1d5110117e4afa7/python/lsst/obs/decam/makeDecamRawVisitInfo.py#L68) to make it set an appropriate pressure, then it seems like fixing it on this ticket would be trivial and we should just do it.

If it's redesigning `VisitInfo`, updating obs packages with new singletons, holding extensive design discussions, etc, then I agree with @isullivan that it belongs on another ticket.